src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/JsonRequestHelpers.cs

```diff
@@ -265,11 +278,60 @@ public static async Task WriteResponseMessage(HttpResponse response, Encoding en
             GrpcServerLog.ReceivedMessage(serverCallContext.Logger);
             return (TRequest)requestMessage;
         }
+        catch (JsonException ex)
+        {
+            GrpcServerLog.ErrorReadingMessage(serverCallContext.Logger, ex);
+            throw new RpcException(new Status(StatusCode.InvalidArgument, "Request JSON payload is not correctly formatted.", ex));
+        }
         catch (Exception ex)
         {
             GrpcServerLog.ErrorReadingMessage(serverCallContext.Logger, ex);
-            throw;
+            throw new RpcException(new Status(StatusCode.InvalidArgument, ex.Message, ex));
+        }
+    }
+
+    private static async ValueTask<IMessage> ReadHttpBodyAsync(JsonTranscodingServerCallContext serverCallContext)
+    {
+        var httpBody = (IMessage)Activator.CreateInstance(serverCallContext.DescriptorInfo.BodyDescriptor!.ClrType)!;
+
+        var contentType = serverCallContext.HttpContext.Request.ContentType;
+        if (contentType != null)
+        {
+            httpBody.Descriptor.Fields[HttpBody.ContentTypeFieldNumber].Accessor.SetValue(httpBody, contentType);
         }
+
+        var data = await ReadDataAsync(serverCallContext);
+        httpBody.Descriptor.Fields[HttpBody.DataFieldNumber].Accessor.SetValue(httpBody, UnsafeByteOperations.UnsafeWrap(data));
+
+        return httpBody;
+    }
+
+    private static async ValueTask<byte[]> ReadDataAsync(JsonTranscodingServerCallContext serverCallContext)
+    {
+        // Buffer to disk if content is larger than 30Kb.
+        // Based on value in XmlSerializer and NewtonsoftJson input formatters.
+        const int DefaultMemoryThreshold = 1024 * 30;
+
+        var memoryThreshold = DefaultMemoryThreshold;
+        var contentLength = serverCallContext.HttpContext.Request.ContentLength.GetValueOrDefault();
+        if (contentLength > 0 && contentLength < memoryThreshold)
+        {
+            // If the Content-Length is known and is smaller than the default buffer size, use it.
+            memoryThreshold = (int)contentLength;
```

Comment:
I'm still not sure this is a good idea for slow-uploading clients. While it's true that this will ultimately be buffered into memory all at once after the request body is fully uploaded, that could take a long time. What do you think @Tratcher?

Reply:
Never mind. I missed three lines up where this is limited by `DefaultMemoryThreshold`.